RecoLocalTracker/SiStripClusterizer/plugins/SiStripClusters2ApproxClusters.cc

```diff
@@ -27,25 +44,114 @@ class SiStripClusters2ApproxClusters : public edm::stream::EDProducer<> {
+
+  fileInPath = edm::FileInPath(SiStripDetInfoFileReader::kDefaultFile);
+  detInfo = SiStripDetInfoFileReader::read(fileInPath.fullPath());
+
+  csfLabel_ = conf.getParameter<std::string>("clusterShapeHitFilterLabel");
+  csfToken_ = esConsumes(edm::ESInputTag("", csfLabel_));
+
+  stripNoiseToken_ = esConsumes();
+
   produces<edmNew::DetSetVector<SiStripApproximateCluster> >();
 }
 
-void SiStripClusters2ApproxClusters::produce(edm::Event& event, edm::EventSetup const&) {
+void SiStripClusters2ApproxClusters::produce(edm::Event& event, edm::EventSetup const& iSetup) {
   auto result = std::make_unique<edmNew::DetSetVector<SiStripApproximateCluster> >();
   const auto& clusterCollection = event.get(clusterToken);
 
+  edm::Handle<reco::BeamSpot> beamSpotHandle;
+  event.getByToken(beamSpotToken, beamSpotHandle);  // retrive BeamSpot data
+  reco::BeamSpot const* bs = nullptr;
+  if (beamSpotHandle.isValid()) {
+    bs = &(*beamSpotHandle);
+  } else {
+    edm::LogError("SiStripClusters2ApproxClusters")
+        << "didn't find a valid beamspot with label " << beamSpot.label() << " using 0,0,0";
+    bs = new reco::BeamSpot();
+  }
+
+  const auto& tkGeom = &iSetup.getData(tkGeomToken_);
+  const auto& theFilter = &iSetup.getData(csfToken_);
+  const auto& theNoise = &iSetup.getData(stripNoiseToken_);
+
   for (const auto& detClusters : clusterCollection) {
     edmNew::DetSetVector<SiStripApproximateCluster>::FastFiller ff{*result, detClusters.id()};
 
-    for (const auto& cluster : detClusters)
-      ff.push_back(SiStripApproximateCluster(cluster, maxNSat));
+    unsigned int detId = detClusters.id();
+    const GeomDet* det = tkGeom->idToDet(detId);
+    double nApvs = detInfo.getNumberOfApvsAndStripLength(detId).first;
+    double stripLength = detInfo.getNumberOfApvsAndStripLength(detId).second;
+    double barycenter_ypos = 0.5 * stripLength;
+
+    const StripGeomDetUnit* stripDet = dynamic_cast<const StripGeomDetUnit*>(det);
+    float mip = 3.9 / (MeVperADCStrip / stripDet->surface().bounds().thickness());
+
+    for (const auto& cluster : detClusters) {
+      const LocalPoint& lp = LocalPoint(((cluster.barycenter() * 10 / (sistrip::STRIPS_PER_APV * nApvs)) -
+                                         ((stripDet->surface().bounds().width()) * 0.5f)),
+                                        barycenter_ypos - (0.5f * stripLength),
+                                        0.);
+      const GlobalPoint& gpos = det->surface().toGlobal(lp);
+      GlobalPoint beamspot(bs->position().x(), bs->position().y(), bs->position().z());
+      const GlobalVector& gdir = gpos - beamspot;
+      const LocalVector& ldir = det->toLocal(gdir);
+
+      int hitStrips;
+      float hitPredPos;
+      theFilter->getSizes(detId, cluster, lp, ldir, hitStrips, hitPredPos);
+
+      bool peakFilter = false;
+      SlidingPeakFinder pf(std::max<int>(2, std::ceil(std::abs(hitPredPos) + subclusterWindow_)));
+      float mipnorm = mip / std::abs(ldir.z());
```

Comment:
Could `ldir.z()` ever be zero ?

Reply:
unless we have the beamspot sitting on one of the strip modules (which I would say it's an unlikely event) I don't think it can be zero.